src/vm/functions/special.rs

```diff
@@ -131,3 +139,173 @@ pub fn handle_contract_call_special_cases(
     // TODO: insert more special cases here, as needed
     Ok(())
 }
+
+/// Given two microblock headers, were they signed by the same key?
+/// Return the pubkey hash if so; return Err otherwise
+fn check_microblock_header_signer(
+    mblock_hdr_1: &StacksMicroblockHeader,
+    mblock_hdr_2: &StacksMicroblockHeader,
+) -> Result<Hash160> {
+    let pkh1 = mblock_hdr_1.check_recover_pubkey().map_err(|e| {
+        InterpreterError::InvalidPoisonMicroblockTransaction(format!(
+            "Failed to recover public key: {:?}",
+            &e
+        ))
+    })?;
+
+    let pkh2 = mblock_hdr_2.check_recover_pubkey().map_err(|e| {
+        InterpreterError::InvalidPoisonMicroblockTransaction(format!(
+            "Failed to recover public key: {:?}",
+            &e
+        ))
+    })?;
+
+    if pkh1 != pkh2 {
+        let msg = format!(
+            "Invalid PoisonMicroblock transaction -- signature pubkey hash {} != {}",
+            &pkh1, &pkh2
+        );
+        warn!("{}", &msg);
+        return Err(InterpreterError::InvalidPoisonMicroblockTransaction(msg).into());
+    }
+    Ok(pkh1)
+}
+
+/// Process a poison-microblock transaction.  This isn't something that can happen from a Clarity
+/// program, but it's "special" because it does a non-trivial amount of VM work that needs to be
+/// accounted for in the encapsulating Environment.
+/// Returns a Value that represents the miner slashed:
+/// * contains the block height of the block with the slashed microblock public key hash
+/// * contains the microblock public key hash
+/// * contains the sender that reported the poison-microblock
+/// * contains the sequence number at which the fork occured
+pub fn handle_poison_microblock(
```

Comment:
I'm not sure if this file/module is the best place for this function. The `vm::functions` module is meant to just be where the Clarity native functions are defined -- this probably makes more sense somewhere in the `chainstate::stacks` module, even though it operates on a lot of Clarity structures.

Reply:
I'll go ahead and put it where I had it originally -- in `chainstate::stacks::db::transactions`.

Reply:
Addressed in ee12847